applications/frontier/src/frontier_handle_rate.erl

```diff
@@ -0,0 +1,329 @@
+fetch_from_parents(AccountDb, MethodList, Realm) ->
+    AccountId = wh_util:format_account_id(AccountDb, 'raw'),
+    case couch_mgr:open_cache_doc(AccountDb, AccountId) of
+        {'ok', JObj} ->
+            Tree = lists:reverse(wh_json:get_value(<<"pvt_tree">>, JObj)),
+            check_fallbacks(Tree, MethodList, Realm);
+        {'error', _Reason} ->
+            lager:info("Cant't access to db: ~p", [_Reason])
+    end.
+
+-spec check_fallbacks(ne_binaries(), ne_binaries(), ne_binary()) -> wh_json:objects().
+check_fallbacks(Tree, MethodList, Realm) ->
+    Result = lists:foldl(fun (X, Acc) -> check_fallback(X, Acc, MethodList, Realm) end, 'empty', Tree),
+    case Result of
+        'empty' -> [];
+        _ -> Result
+    end.
+
+-spec check_fallback(ne_binary(), atom() | wh_json:objects(), ne_binaries(), ne_binary()) -> atom() | wh_json:objects().
+check_fallback(AccountId, 'empty', MethodList, Realm) ->
+    AccountDB = wh_util:format_account_id(AccountId, 'encoded'),
+    ViewOpts = [{'key', AccountId}],
+    case couch_mgr:get_results(AccountDB, <<"rate_limits/list_by_owner">>, ViewOpts) of
+        {'ok', []} -> 'empty';
+        {'ok', [JObj]} ->
+            Fallback = wh_json:get_value(<<"value">>, JObj),
+            build_results(couch_mgr:open_cache_doc(AccountDB, Fallback), MethodList, Realm);
+        {'ok', _JObjs} ->
+            lager:error("found many results, please check account rate limits for ~s", [AccountDB]),
+            'empty';
+        {'error', _Reason} ->
+            lager:error("Can't fetch data from db: ~p", [_Reason]),
+            'empty'
+    end;
+check_fallback(_, Acc, _, _) ->
+    Acc.
+
+-type couch_ret() :: {'ok', wh_json:object()} | {'error', any()}.
+-spec build_results(couch_ret(), ne_binaries(), ne_binary()) -> wh_json:objects().
+build_results({'error', _Reason}, _, _) ->
+    lager:error("Can't fetch data from db: ~p", [_Reason]),
+    [];
+build_results({'ok', JObj}, MethodList, Realm) ->
+    Limits = wh_json:get_value(<<"account">>, JObj),
+    lists:foldl(fun (M, Acc) -> build(M, Acc, Limits, Realm) end,[],MethodList).
+
+-spec build(ne_binary(), wh_json:objects(), wh_json:object(), ne_binary()) -> wh_json:objects().
+build(Method, Acc, JObj, Realm) ->
+    PerMinute = wh_json:get_value([?MINUTE, Method], JObj),
+    PerSecond = wh_json:get_value([?SECOND, Method], JObj),
+    case PerMinute =:= 'undefined' orelse PerSecond =:= 'undefined' of
+        'true' -> Acc;
+        'false' -> construct_records(Method, Realm, PerMinute, PerSecond)
+    end.
+
+-type status() :: 'ok' | 'need_account'.
+-type rates_ret() :: {status(), wh_json:objects()}.
+-spec handle_db_response(wh_json:objects(), boolean()) -> rates_ret().
+handle_db_response(JObjs, IncludeRealm) ->
+    {DefaultRates, OtherRates} = lists:partition(fun is_device_defaults/1, JObjs),
+    {AccountRates, DeviceRates} = lists:partition(fun frontier_utils:is_realm/1, OtherRates),
+    DeviceResult = case length(DeviceRates) > 0 of
```

Comment:
Getting the length of a list just to see if it is empty or not isn't a great thing to do, in general. I would change this to be more like:

```erlang
case DeviceRates of
    [] -> DefaultRates;
    _ -> DeviceRates
end
```
Still testing emptiness, but without having to count the length of the list (afaik Erlang doesn't track the length and has to compute it when `length/1` is called).